packages/paste-core/components/chat-log/src/useChatLog.ts

```diff
@@ -0,0 +1,61 @@
+import * as React from 'react';
+import {uid} from '@twilio-paste/uid-library';
+
+type ChatDictionary = Record<string, ChatData>;
+type PushChat = (chat: ChatData) => void;
+type PopChat = (id: string) => void;
+type Chats = React.ReactElement[];
+
+export type ChatData = {id?: string; content: React.ReactElement};
+export type UseChatLog = (...initialChats: ChatData[]) => {
+  chats: Chats;
+  chatIds: string[];
+  push: PushChat;
+  pop: PopChat;
+};
+
+const spreadChatIntoDictionary = (prev: ChatDictionary, next: ChatData): ChatDictionary => ({
+  ...prev,
+  [next.id || uid(next.content)]: next,
+});
+
+export const useChatLog: UseChatLog = (...initialChats) => {
+  const parsedInitialChats = React.useMemo(
+    () => initialChats.reduce((prev, next) => spreadChatIntoDictionary(prev, next), {}),
+    [initialChats]
+  );
+
+  const [chatDictionary, setChatDictionary] = React.useState<ChatDictionary>(parsedInitialChats);
+
+  const push: PushChat = React.useCallback(
+    (next) => {
+      setChatDictionary((prev) => spreadChatIntoDictionary(prev, next));
+    },
+    [setChatDictionary]
+  );
+
+  const pop: PopChat = React.useCallback(
+    (id) => {
+      setChatDictionary((prev) => {
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        const {[id]: _, ...restPrev} = prev;
+        return restPrev;
+      });
+    },
+    [setChatDictionary]
+  );
+
+  const parsedChatData = React.useMemo(
+    () =>
+      Object.entries(chatDictionary).reduce<{chats: Chats; chatIds: string[]}>(
+        (prev, [id, {content}]) => ({
+          chats: prev.chats.concat(React.cloneElement(content, {key: id})),
```

Comment:
Thought: again here, where we're cloning children is another example where the API using jsx might lend itself too much to overreaching into things we don't own. We're cloning and inspecting the children to make this work and it's not a pattern we tend to use due to component encapsulation rules. I.e no component should know about the existance of other components, but it relies on an implicit contract that could change if the consumer changes the child